packages_rs/nextclade/src/tree/tree_builder.rs

```diff
@@ -193,6 +193,12 @@ pub fn finetune_nearest_node(
           current_best_node.payload().name
         )
       })?;
+      private_mutations = union_of_muts(&private_mutations, &best_split_result.left.invert()).wrap_err_with(|| {
+        format!(
+          "When calculating union of mutations between query sequence the zero-length parent node '{}'",
+          graph.get_node(best_node.key()).expect("Node not found").payload().name
+        )
+      })?;
       current_best_node = graph
         .parent_of_by_key(best_node.key())
         .ok_or_else(|| make_internal_report!("Parent node is expected, but not found"))?;
```

Comment:
In terms of science:

 - Is the order of set operations significant here, and is it correct? I believe that `union(diff(pm, shared), left)` and `diff(union(pm, left), shared)` might yield different results.


A few technicalities:

 - The error message seems to be missing a word ("sequence and ..."?) or otherwise incomplete. It is also a bit asymmetric with the one from the diff operation.
 - I believe that the node name can be retrieved simpler, like in the diff operation. Unless it's a different name. But I think right now it just makes a round trip: node -> node key ->  (fallible operation) -> node -> payload -> name, instead of just node -> payload -> name.
 - Might need some more explanatory comments - the code flow in this function is so complex and poorly structured that I cannot understand a thing anymore, just a few weeks after its written. After a few bug fixes added, it will become gibberish even for you :) 


A few unrelated ideas for the future of this function (I am writing down to not forget):

 - I hope we can refactor control flow in this loop sometimes too.
 - In a single-threaded graph implementation perhaps generic `.payload()` makes less sense. Intrusive storage right in the node struct might make code more readable (shorter chains to access things).
 - The parallel variables: best_node, best_split_result, n_shared_muts - they are flowing together and are modified together. This is verbose and error-pron (can forget to modify one of them). Might be possible to aggregate them into one variable.
 - Lambda functions in error handling can be extracted into named functions to make the messages less intrusive.